planner/cascades/implementation_rules.go

```diff
@@ -154,22 +158,71 @@ func (r *ImplTiKVSingleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *pr
+func (r *ImplTiKVDoubleReadGather) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	if prop.IsEmpty() {
+		return true
+	}
+	reader := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	if all, _ := prop.AllSameOrder(); all {
+		for i, col := range reader.IndexCols {
+			if col.Equal(nil, prop.Items[0].Col) {
+				return plannercore.MatchIndicesProp(reader.IndexCols[i:], reader.IndexColLens[i:], prop.Items)
+			}
+		}
+	}
+	return false
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplTiKVDoubleReadGather) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	logicProp := expr.Group.Prop
+	dg := expr.ExprNode.(*plannercore.TiKVDoubleGather)
+	var reader plannercore.PhysicalPlan
+	var proj *plannercore.PhysicalProjection
+	reader = dg.GetPhysicalIndexLookUpReader(logicProp.Schema, logicProp.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt), reqProp.Clone(), reqProp.Clone())
+	if dg.HandleCol.ID == model.ExtraHandleID && !reqProp.IsEmpty() {
+		// TODO: if the origin schema for IndexLookUpReader has the `HandleCol`, we can remove the duplicated append.
+		reader.Schema().Append(dg.HandleCol)
+		reader.(*plannercore.PhysicalIndexLookUpReader).ExtraHandleCol = dg.HandleCol
+		proj = plannercore.PhysicalProjection{Exprs: expression.Column2Exprs(logicProp.Schema.Columns)}.Init(dg.SCtx(), logicProp.Stats, dg.SelectBlockOffset(), reqProp.Clone(), reqProp.Clone())
+		proj.SetSchema(logicProp.Schema)
+		proj.SetChildren(reader)
+	}
+	indexLookUp := impl.NewIndexLookUpReaderImpl(reader, dg.Source.TblColHists, proj)
+	indexLookUp.KeepOrder = !reqProp.IsEmpty()
+	return indexLookUp, nil
+}
+
 // ImplTableScan implements TableScan as PhysicalTableScan.
 type ImplTableScan struct {
 }
 
 // Match implements ImplementationRule Match interface.
 func (r *ImplTableScan) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
 	ts := expr.ExprNode.(*plannercore.LogicalTableScan)
-	return prop.IsEmpty() || (len(prop.Items) == 1 && ts.Handle != nil && prop.Items[0].Col.Equal(nil, ts.Handle))
+	return prop.IsEmpty() || ts.IsDoubleRead || (len(prop.Items) == 1 && ts.Handle != nil && prop.Items[0].Col.Equal(nil, ts.Handle))
 }
 
 // OnImplement implements ImplementationRule OnImplement interface.
 func (r *ImplTableScan) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
 	logicProp := expr.Group.Prop
 	logicalScan := expr.ExprNode.(*plannercore.LogicalTableScan)
 	ts := logicalScan.GetPhysicalScan(logicProp.Schema, logicProp.Stats.ScaleByExpectCnt(reqProp.ExpectedCnt))
-	if !reqProp.IsEmpty() {
+	if logicalScan.IsDoubleRead {
+		if !reqProp.IsEmpty() {
+			if logicalScan.Source.HandleCol == nil {
+				ts.AppendExtraHandleCol(logicalScan.Source)
+			} else if logicalScan.Source.HandleCol.ID == model.ExtraHandleID {
+				ts.Schema().Append(logicalScan.Source.HandleCol)
```

Comment:
Here we should check whether the schema contains handle column actually?
The handle col may not be pruned if it appeared in the select field.

Reply:
Done.